### PR TITLE
created tests for CalculatorImpl

### PR DESCRIPTION
**Pull request by Daniel1004K**

### CODE REVIEW

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            return subtract(a, b);
+        }
+        if (operation == '/') {
+            return div(a, b);
+        }
+        if (operation == '^') {
+            return pow(a, b);
+        }
+        throw new IllegalArgumentException("Not defined function");
+    }
```
Comment (Tamara-Paskalova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:23]: In my opinion it's better to make the values of operation characters constant.

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            for (int j = 0; j < 10; j++) {
+                assertEquals(i + j, calculator.calculate(i, j, '+'));
+            }
+        }
+        double a = 0.5;
+        double b = 0.345;
+        double expected = a + b;
+        double actual = calculator.calculate(a, b, '+');
+        assertEquals(expected, actual);
+    }
```
Comment (Tamara-Paskalova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:30]: In my opinion last part of this test is quite enough. The loop is redundant

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+    @Test
+    void calculate_illegalOperation_notOk() {
+        assertThrows(IllegalArgumentException.class, () -> {
+            for (int i = 0; i < 1000; i++) {
+                if (i == (int) '+' || i == (int) '-' || i == (int) '/' || i == (int) '*'
+                        || i == (int) '^') {
+                    continue;
+                }
+                calculator.calculate(34, 53, (char) i);
```
Comment (Tamara-Paskalova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:291]: It's hard for me to understand what are you doing here

File: src/main/java/core/basesyntax/CalculatorImpl.java (hunk 1 above)
Comment (Daniel1004K) [reply src/main/java/core/basesyntax/CalculatorImpl.java:23]: I think so

**Review by Tamara-Paskalova — APPROVED** [review]

**Review by Tamara-Paskalova — APPROVED** [review]
Good job!

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package core.basesyntax;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CalculatorImpl implements Calculator {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:7]: redundant empty line

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    public static final String NUMBERS = "-?\\d+\\.?\\d+|-?\\d+";
+    public static final String CHARACTER = "[+/*^]|--";
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:9]:
```suggestion
    public static final String NUMBER_REGEX = "-?\\d+\\.?\\d+|-?\\d+";
    public static final String CHARACTER_REGEX = "[+/*^]|--";
```

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        if (operation == '-') {
+            return subtract(a, b);
+        }
+        if (operation == '/') {
+            return div(a, b);
+        }
+        if (operation == '^') {
+            return pow(a, b);
+        }
+        throw new IllegalArgumentException("Not defined function");
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:38]: let's better use switch case

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        return a + b;
+    }
+
+    @Override
+    public double subtract(double a, double b) {
+        return a - b;
+    }
+
+    @Override
+    public double div(double a, double b) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:52]:
```suggestion
    public double divide(double a, double b) {
```

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        return a / b;
+    }
+
+    @Override
+    public double multiply(double a, double b) {
+        return a * b;
+    }
+
+    @Override
+    public double pow(double a, double power) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:65]:
```suggestion
    public double power(double a, double power) {
```

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        return Math.pow(a, power);
+    }
+
+    private void checkExpression(String expression) {
+        if (!expression.matches("-?\\d+\\.?\\d+\\s?[+*^/]\\s*-?\\d+\\.?\\d+|"
+                + "-?\\d+\\s?[+*^/]\\s*-?\\d+|"
+                + "-?\\d+\\s?--\\s*-?\\d+|"
+                + "-?\\d+\\.?\\d+\\s?--\\s*-?\\d+\\.?\\d+")) {
+            throw new IllegalArgumentException();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:74]: let's add informative message here

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    private char getCharacter(Matcher matcher) {
+        if (matcher.find()) {
+            return matcher.group().charAt(0);
+        }
+        throw new IllegalArgumentException();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:82]: same

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    private double getNum(Matcher matcher) {
+        if (matcher.find()) {
+            return Double.parseDouble(matcher.group());
+        }
+        throw new IllegalArgumentException();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/CalculatorImpl.java:89]: same

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+import org.junit.jupiter.api.Test;
+
+class CalculatorImplTest {
+    public static final char PLUS = '+';
+    public static final char MINUS = '-';
+    public static final char CHARACTER_MULTIPLICATION = '*';
+    public static final char CHARACTER_DIVISION = '/';
+    public static final char CHARACTER_POWER = '^';
+    private static CalculatorImpl calculator;
+    private static final double DELTA = 0.0001;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:16]: let's switch these 2 lines - constants should be declared above other fields

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+
+    @Test
+    public void calculate_ExpressionTrue_notOk() {
+        assertThrows(IllegalArgumentException.class, () -> {
+            for (int character = 0; character < 1000; character++) {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:21]: this seems to take the first available char that is not suitable, call calculate(), get the exception, and it is over - we won't go further in the loop

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+        assertEquals(-35.56 + -30.56, calculator.calculate(-35.56 + " + " + -30.56), DELTA);
+        assertEquals(35.56 + -30.56, calculator.calculate(35.56 + " + " + -30.56), DELTA);
+        assertEquals(35.56 + 30.56, calculator.calculate(35.56 + " + " + 30.56), DELTA);
+    }
+
+    @Test
+    public void calculate_subtractionExpression_ok() {
+        for (int i = -100; i < 100; i++) {
+            for (int j = -100; j < 100; j++) {
+                assertEquals(i - j, calculator.calculate(i + " -- " + j), DELTA,
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:51]: why `--`?

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+                        "i = " + i + " j = " + j);
+            }
+        }
+        assertEquals(-35.56 - 30.56, calculator.calculate(-35.56 + " -- " + 30.56), DELTA);
+        assertEquals(-35.56 - -30.56, calculator.calculate(-35.56 + " -- " + -30.56), DELTA);
+        assertEquals(35.56 - -30.56, calculator.calculate(35.56 + " -- " + -30.56), DELTA);
+        assertEquals(35.56 - 30.56, calculator.calculate(35.56 + " -- " + 30.56), DELTA);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:58]: same

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+        assertEquals(35.56 / 30.56, calculator.calculate(35.56 + " / " + 30.56), DELTA);
+    }
+
+    @Test
+    public void calculate_multiplyExpression_ok() {
+        for (int i = -100; i < 100; i++) {
+            for (int j = -100; j < 100; j++) {
+                if (j == 0) {
+                    continue;
+                }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:84]: why do we skip 0?

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+                assertEquals((double) i * j, calculator.calculate(i + " * " + j), DELTA,
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:85]:
```suggestion
                assertEquals(i * j, calculator.calculate(i + " * " + j), DELTA,
```

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+        assertEquals(35.56 * 30.56, calculator.calculate(35.56 + " * " + 30.56), DELTA);
+    }
+
+    @Test
+    public void calculate_powerExpression_ok() {
+        for (int i = -100; i < 100; i++) {
+            for (int j = -100; j < 100; j++) {
+                if (j == 0) {
+                    continue;
+                }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:101]: why do we skip 0?

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+        assertEquals(Math.pow(35.56, -30.56), calculator.calculate(35.56 + " ^ " + -30.56), DELTA);
+        assertEquals(Math.pow(35.56, 30.56), calculator.calculate(35.56 + " ^ " + 30.56), DELTA);
+        assertEquals(Math.pow(-35.56, -30.56), calculator.calculate(-35.56 + " ^ " + -30.56),
+                DELTA);
+    }
+
+    @BeforeAll
+    static void setCalculator() {
+        calculator = new CalculatorImpl();
+    }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:116]: please move this method to the beginning of the class

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+        for (int i = 100; i > 0; i--) {
+            for (int j = -100; j < 0; j++) {
+                assertEquals(i + j, calculator.calculate(i, j, PLUS));
+            }
+        }
+        double a = -0.5;
+        double b = 0.345;
+        double expected = a + b;
+        double actual = calculator.calculate(a, b, PLUS
+        );
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:157]:
```suggestion
        double actual = calculator.calculate(a, b, PLUS);
```

File: src/test/java/core/basesyntax/CalculatorImplTest.java [hunk 21]
```diff
@@ -0,0 +1,400 @@
...
+    @Test
+    void calculate_raisingZeroToPower_ok() {
+        double expected = 0;
+        double actual = calculator.calculate(0, 35, CHARACTER_POWER);
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void calculate_illegalOperation_notOk() {
+        assertThrows(IllegalArgumentException.class, () -> {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorImplTest.java:386]: this seems to take the first available char that is not suitable, call calculate(), get the exception, and it is over - we won't go further in the loop

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
an interesting solution, nice job!

**Review by Sofasmile — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/CalculatorImpl.java [hunk 22]
```diff
@@ -84,39 +86,44 @@ private void checkExpression(String expression) {
                 "-?(\\d*\\.?\\d*)\\s*[+*^/-]\\s*-?(\\d*\\.?\\d*)")) {
             return;
         }
-        throw new IllegalArgumentException();
+        throw new IllegalArgumentException("Input expression is wrong.");
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/CalculatorImpl.java:89]:
```suggestion
        throw new IllegalArgumentException("Input expression - " + expression +" is wrong.");
```
same everywhere

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**